### PR TITLE
Put a buffer of 100 logs between log parser and log bundler. Publish …

### PR DESCRIPTION
**Pull request by gkodali-zededa**
…average logs send API time for device and app bundles.

Signed-off-by: Gopi krishna Kodali <gkodali@zededa.com>

### CODE REVIEW

File: pkg/pillar/cmd/logmanager/logmanager.go [hunk 1]
```diff
@@ -738,13 +744,7 @@ func handleAppLogEvent(event logEntry, appLogs *logs.AppInstanceLogBundle) bool
...
 
 	logDetails := &logs.LogEntry{}
-	// XXX Is this still required. rsyslogd is now configured to do the same
-	logDetails.Content = strings.Map(func(r rune) rune {
-		if r == utf8.RuneError {
-			return -1
-		}
-		return r
-	}, event.content)
+	logDetails.Content = event.content
```
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go:747]: How do we know this isn't required? Have you inserted some invalid UTF-8 strings in the dmesg output or something else which feeds into rsyslogd?

File: pkg/pillar/cmd/logmanager/logmanager.go [hunk 2]
```diff
@@ -803,9 +798,9 @@ func handleLogEvent(event logEntry, reportLogs *logs.LogBundle) bool {
 
 // Returns true if a message was successfully sent
```
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go:799]: Please expand the comment to specify the new return

File: pkg/pillar/cmd/logmanager/logmanager.go [hunk 3]
```diff
@@ -868,18 +865,20 @@ func sendProtoStrForLogs(reportLogs *logs.LogBundle, image string,
...
+		return false, uint64(totalBundleSend / time.Millisecond)
 	}
 	log.Debugf("Sent %d bytes image %s to %s", size, image, logsURL)
 	reportLogs.Log = []*logs.LogEntry{}
-	return true
+	totalBundleSend := time.Since(startOfBundleSend)
+	return true, uint64(totalBundleSend / time.Millisecond)
 }
 
 // Returns true if a message was successfully sent
```
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go:877]: Ditto

File: api/proto/metrics/metrics.proto [hunk 4]
```diff
@@ -228,6 +228,9 @@ message logMetric {
 
   uint64 numDeviceBundleProtoBytesSent = 17;
   uint64 numAppBundleProtoBytesSent = 18;
+
+  uint64 avgDeviceBundleSendTimeMillis = 19; // protolint:disable:this FIELD_NAMES_LOWER_SNAKE_CASE
+  uint64 avgAppBundleSendTimeMillis = 20; // protolint:disable:this FIELD_NAMES_LOWER_SNAKE_CASE
```
Comment (eriknordmark) [inline api/proto/metrics/metrics.proto:233]: These protolints didn't silence prototool.

But you can define them in lower snake case and the generated fields will be CamelCase.

**Review by eriknordmark — COMMENTED** [review]
Do we want this time measurement in the product?
If so should we just measure logmanager or also zedagent etc?

**Comment by rvs** [comment]
Hey @eriknordmark 

> Do we want this time measurement in the product?

What do you mean by "the product"?

> If so should we just measure logmanager or also zedagent etc?

I think that would actually make a lot of sense

File: api/proto/metrics/metrics.proto (hunk 4 above)
Comment (gkodali-zededa) [reply api/proto/metrics/metrics.proto:233]: Ok. I'll change these two members to lower snake case.

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 2 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go:799]: will do

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 3 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go:877]: sure

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 1 above)
Comment (gkodali-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go:747]: rsyslogd as mmutf8fix module enables. This module replaces any invalid utf-8 sequences with a space.

Tested by sending a valid utf8
- logger $'\xe2\x80\xa2'

Tested by sending invalid utf8
- logger $'\xe2\x80\xa2'
I saw this invalid sequence replaced with space by rsyslogd.

**Comment by gkodali-zededa** [comment]
> Do we want this time measurement in the product?
> If so should we just measure logmanager or also zedagent etc?

Should we put this API call average measurement as part of zedcloud metrics?
We can add a per API/URL per agent metric that includes the number of calls, average time spent waiting for result & timestamp of last call etc

**Comment by eriknordmark** [comment]
> What do you mean by "the product"?
> 
> >

Meaning that it will be a permanent addition as opposed to temporary while we debug and tune the logging.

**Comment by eriknordmark** [comment]
> Should we put this API call average measurement as part of zedcloud metrics?
> We can add a per API/URL per agent metric that includes the number of calls, average time spent waiting for result & timestamp of last call etc

We already have the number of calls plus LastSucceeded/LastFailed. Adding the total time for all the calls (the user can compute averages) would make sense.
Don't know if we want to track the time for failed calls separately from succeeded calls.

**Comment by gkodali-zededa** [comment]
> > Should we put this API call average measurement as part of zedcloud metrics?
> > We can add a per API/URL per agent metric that includes the number of calls, average time spent waiting for result & timestamp of last call etc
> 
> We already have the number of calls plus LastSucceeded/LastFailed. Adding the total time for all the calls (the user can compute averages) would make sense.
> Don't know if we want to track the time for failed calls separately from succeeded calls.

Will add this code.

**Comment by gkodali-zededa** [comment]
@rvs @nordmark Can you please take a look now?

File: api/proto/metrics/metrics.proto [hunk 5]
```diff
@@ -70,6 +70,7 @@ message urlcloudMetric {
   int64 sentByteCount = 5;
   int64 recvMsgCount = 6;
   int64 recvByteCount = 7;
+  int64 total_time_spent = 8;
```
Comment (eriknordmark) [inline api/proto/metrics/metrics.proto:73]: For successes and failures or just successes? Please specify in a comment.

File: pkg/pillar/cmd/logmanager/logmanager.go [hunk 6]
```diff
@@ -613,20 +612,24 @@ func processEvents(image string, logChan <-chan logEntry,
...
 					logMetrics.LastAppBundleSendTime = time.Now()
 				}
 				delete(appLogBundles, appUUID)
 			} else {
-				sent = sendProtoStrForLogs(reportLogs, image, iteration, eveVersion)
+				var devSendTime uint64
+				sent, devSendTime = sendProtoStrForLogs(reportLogs, image, iteration, eveVersion)
 				logMetrics.NumDeviceEventsSent += uint64(messageCount)
 				logMetrics.NumDeviceBundleProtoBytesSent += uint64(byteCount)
+				logMetrics.AvgDeviceBundleSendTime = (logMetrics.AvgDeviceBundleSendTime*logMetrics.NumDeviceBundlesSent + devSendTime) / (logMetrics.NumDeviceBundlesSent + 1)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go:632]: 1. Do we still need this as part of LogMetrics given that we have it in zedcloudMetrics?
2. Why re-compute an average every time? Tracking the sum and count is sufficient.

File: pkg/pillar/cmd/logmanager/logmanager.go (hunk 1 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/logmanager/logmanager.go:747]: OK

**Review by eriknordmark — COMMENTED** [review]
There are 4 or 5 different fixes in here and the description and commit messages do not capture that.
I suspect we could make faster progress if there were separate smaller PRs.

**Comment by gkodali-zededa** [comment]
Erik I made some changes in the same area and that has made it almost impossible for me to rebase.
I'll close this PR and separate new PRs.

**Pull request closed without merging**